plugins/modules/nd_setup.py

```diff
@@ -0,0 +1,358 @@
+from ansible_collections.cisco.nd.plugins.module_utils.nd_argument_specs import network_spec, bgp_spec
+
+
+def main():
+    argument_spec = nd_argument_spec()
+    argument_spec.update(
+        cluster_name=dict(type="str"),
+        ntp_server=dict(type="list", aliases=["ntp_servers"], elements="str"),
+        dns_server=dict(type="list", aliases=["dns_servers"], elements="str"),
+        proxy_server=dict(type="str"),
+        proxy_username=dict(type="str"),
+        proxy_password=dict(type="str", no_log=True),
+        ignore_proxy=dict(type="list", elements="str"),
+        dns_search_domain=dict(type="list", aliases=["dns_search_domains"], elements="str"),
+        app_network=dict(type="str"),
+        service_network=dict(type="str"),
+        nodes=dict(
+            type="list",
+            elements="dict",
+            options=dict(
+                hostname=dict(type="str", required=True),
+                serial_number=dict(type="str", required=True),
+                management_ip_address=dict(type="str", required=True),
+                username=dict(type="str", required=True),
+                password=dict(type="str", required=True, no_log=True),
+                management_network=dict(type="dict", required=True, options=network_spec()),
+                data_network=dict(type="dict", required=True, options=network_spec(vlan=True)),
+                bgp=dict(type="dict", options=bgp_spec()),
+            ),
+        ),
+        state=dict(type="str", default="present", choices=["present", "query"]),
+    )
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        required_if=[
+            ["state", "present", ["cluster_name", "ntp_server", "dns_server", "app_network", "service_network", "nodes"]],
+        ],
+        required_together=[
+            ["proxy_username", "proxy_password"],
+        ],
+    )
+
+    nd = NDModule(module)
+
+    cluster_name = nd.params.get("cluster_name")
+    ntp_server = nd.params.get("ntp_server")
+    dns_server = nd.params.get("dns_server")
+    proxy_server = nd.params.get("proxy_server")
+    proxy_username = nd.params.get("proxy_username")
+    proxy_password = nd.params.get("proxy_password")
+    ignore_proxy = nd.params.get("ignore_proxy")
+    dns_search_domain = nd.params.get("dns_search_domain")
+    app_network = nd.params.get("app_network")
+    service_network = nd.params.get("service_network")
+    nodes = nd.params.get("nodes")
+    state = nd.params.get("state")
+
+    if state == "query":
+        nd.existing = nd.request("/clusterstatus/install", method="GET")
+    else:
+        if len(cluster_name) > 63 or len(re.findall(r"[^a-zA-Z0-9 -]|^-| *-$", cluster_name)) > 0:
+            nd.fail_json(
+                "A length of 1 to 63 characters is allowed and valid characters include letters, digits and hyphen. "
+                "The name cannot start or end with a hyphen."
```

Comment:
should we do two distinct conditions for this so that the user knows exactly what is the error?

Reply:
That makes sense but this is how it is in the UI.